// @antv/infographic/lib/editor/interactions/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoomWheel = exports.SelectHighlight = exports.HotkeyHistory = exports.DragElement = exports.DragCanvas = exports.DblClickEditText = exports.ClickSelect = exports.BrushSelect = exports.Interaction = void 0;
var base_1 = require("./base");
Object.defineProperty(exports, "Interaction", { enumerable: true, get: function () { return base_1.Interaction; } });
var brush_select_1 = require("./brush-select");
Object.defineProperty(exports, "BrushSelect", { enumerable: true, get: function () { return brush_select_1.BrushSelect; } });
var click_select_1 = require("./click-select");
Object.defineProperty(exports, "ClickSelect", { enumerable: true, get: function () { return click_select_1.ClickSelect; } });
var dblclick_edit_text_1 = require("./dblclick-edit-text");
Object.defineProperty(exports, "DblClickEditText", { enumerable: true, get: function () { return dblclick_edit_text_1.DblClickEditText; } });
var drag_canvas_1 = require("./drag-canvas");
Object.defineProperty(exports, "DragCanvas", { enumerable: true, get: function () { return drag_canvas_1.DragCanvas; } });
var drag_element_1 = require("./drag-element");
Object.defineProperty(exports, "DragElement", { enumerable: true, get: function () { return drag_element_1.DragElement; } });
var hotkey_history_1 = require("./hotkey-history");
Object.defineProperty(exports, "HotkeyHistory", { enumerable: true, get: function () { return hotkey_history_1.HotkeyHistory; } });
var select_highlight_1 = require("./select-highlight");
Object.defineProperty(exports, "SelectHighlight", { enumerable: true, get: function () { return select_highlight_1.SelectHighlight; } });
var zoom_wheel_1 = require("./zoom-wheel");
Object.defineProperty(exports, "ZoomWheel", { enumerable: true, get: function () { return zoom_wheel_1.ZoomWheel; } });
