// @antv/infographic/lib/templates/list-zigzag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.listZigzagTemplates = void 0;
exports.listZigzagTemplates = {
    'list-zigzag-up-compact-card': {
        design: {
            title: 'default',
            structure: {
                type: 'list-zigzag-up',
            },
            items: [
                {
                    type: 'compact-card',
                },
            ],
        },
        themeConfig: {
            colorPrimary: '#17CA2C',
        },
    },
    'list-zigzag-up-simple': {
        design: {
            title: 'default',
            structure: {
                type: 'list-zigzag-up',
            },
            items: [
                {
                    type: 'simple',
                    usePaletteColor: true,
                },
            ],
        },
        themeConfig: {
            colorPrimary: '#17CA2C',
        },
    },
    'list-zigzag-down-compact-card': {
        design: {
            title: 'default',
            structure: {
                type: 'list-zigzag-down',
            },
            items: [
                {
                    type: 'compact-card',
                },
            ],
        },
        themeConfig: {
            colorPrimary: '#17CA2C',
        },
    },
    'list-zigzag-down-simple': {
        design: {
            title: 'default',
            structure: {
                type: 'list-zigzag-down',
            },
            items: [
                {
                    type: 'simple',
                    usePaletteColor: true,
                },
            ],
        },
        themeConfig: {
            colorPrimary: '#17CA2C',
        },
    },
};
