// @antv/infographic/lib/renderer/composites/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderText = exports.renderStaticText = exports.renderItemText = exports.renderSVG = exports.renderStaticShape = exports.renderShape = exports.renderItemIllus = exports.renderIllus = exports.renderItemIcon = exports.renderButtonsGroup = exports.renderBaseElement = exports.renderBackground = void 0;
var background_1 = require("./background");
Object.defineProperty(exports, "renderBackground", { enumerable: true, get: function () { return background_1.renderBackground; } });
var base_1 = require("./base");
Object.defineProperty(exports, "renderBaseElement", { enumerable: true, get: function () { return base_1.renderBaseElement; } });
var button_1 = require("./button");
Object.defineProperty(exports, "renderButtonsGroup", { enumerable: true, get: function () { return button_1.renderButtonsGroup; } });
var icon_1 = require("./icon");
Object.defineProperty(exports, "renderItemIcon", { enumerable: true, get: function () { return icon_1.renderItemIcon; } });
var illus_1 = require("./illus");
Object.defineProperty(exports, "renderIllus", { enumerable: true, get: function () { return illus_1.renderIllus; } });
Object.defineProperty(exports, "renderItemIllus", { enumerable: true, get: function () { return illus_1.renderItemIllus; } });
var shape_1 = require("./shape");
Object.defineProperty(exports, "renderShape", { enumerable: true, get: function () { return shape_1.renderShape; } });
Object.defineProperty(exports, "renderStaticShape", { enumerable: true, get: function () { return shape_1.renderStaticShape; } });
var svg_1 = require("./svg");
Object.defineProperty(exports, "renderSVG", { enumerable: true, get: function () { return svg_1.renderSVG; } });
var text_1 = require("./text");
Object.defineProperty(exports, "renderItemText", { enumerable: true, get: function () { return text_1.renderItemText; } });
Object.defineProperty(exports, "renderStaticText", { enumerable: true, get: function () { return text_1.renderStaticText; } });
Object.defineProperty(exports, "renderText", { enumerable: true, get: function () { return text_1.renderText; } });
