// @antv/infographic/esm/index.js
export * from './designs/index.js';
export { getItemProps, getThemeColors } from './designs/utils/index.js';
export { BatchCommand, UpdateElementCommand, UpdateOptionsCommand, UpdateTextCommand, } from './editor/commands/index.js';
export { BrushSelect, ClickSelect, DblClickEditText, DragCanvas, DragElement, HotkeyHistory, Interaction, SelectHighlight, ZoomWheel, } from './editor/interactions/index.js';
export { EditBar, Plugin, ResetViewBox, ResizeElement } from './editor/plugins/index.js';
export { exportToSVG } from './exporter/index.js';
export { Defs, Ellipse, Fragment, Group, Path, Polygon, Rect, Text, cloneElement, createFragment, createLayout, getCombinedBounds, getElementBounds, getElementsBounds, jsx, jsxDEV, jsxs, renderSVG, } from './jsx/index.js';
export { getFont, getFonts, getPalette, getPaletteColor, getPalettes, registerFont, registerPalette, registerPattern, setDefaultFont, } from './renderer/index.js';
export { loadSVGResource, registerResourceLoader } from './resource/index.js';
export { Infographic } from './runtime/index.js';
export { parseSyntax } from './syntax/index.js';
export { getTemplate, getTemplates, registerTemplate } from './templates/index.js';
export { getTheme, getThemes, registerTheme } from './themes/index.js';
export { parseSVG, setFontExtendFactor } from './utils/index.js';
export { VERSION } from './version.js';
