// @antv/infographic/esm/designs/components/Btn.js
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx } from "../../jsx-runtime.js";
import { Rect } from '../../jsx/index.js';
export const BtnAdd = (props) => {
    const { indexes } = props, restProps = __rest(props, ["indexes"]);
    const defaultProps = {
        fill: '#B9EBCA',
        fillOpacity: 0.3,
        width: 20,
        height: 20,
        'data-indexes': indexes,
        'data-element-type': "btn-add" /* ElementTypeEnum.BtnAdd */,
    };
    return _jsx(Rect, Object.assign({}, defaultProps, restProps));
};
export const BtnRemove = (props) => {
    const { indexes } = props, restProps = __rest(props, ["indexes"]);
    const defaultProps = {
        fill: '#F9C0C0',
        fillOpacity: 0.3,
        width: 20,
        height: 20,
        'data-indexes': indexes,
        'data-element-type': "btn-remove" /* ElementTypeEnum.BtnRemove */,
    };
    return _jsx(Rect, Object.assign({}, defaultProps, restProps));
};
